proposed/activity-improvements.md

```diff
@@ -0,0 +1,393 @@
+        public IDictionary<string, object>? Attributes { get; }
+    }
+}
+```
+
+### ActivityCreationOptions
+
+```C#
+namespace System.Diagnostics
+{
+    // ActivityCreationOptions contains values telling if decide to create the Activity object
+    // and with what state
+    // This enum value will be returned from the callback of the Activity listener object which created by ActivitySource.AddActivityListener.
+    public enum ActivityCreationOptions
+    {
+        /// Don't allow creating the Activity object.
+        NoCreate,
+
+        /// Create the Activity but drop recording events and tags.
+        NoDataAllowed,
+
+        /// Create the Activity and allow recording events and tags
+        AllowData,
+
+        /// Create the Activity, allow recording events and tags, and turn on ActivityTraceFlags.Recorded (i.e. Allow Sampling).
+        AllowDataAndSampling
+    }
+}
+```
+
+### ActivitySource
+
+```C#
+
+namespace System.Diagnostics
+{
+    /// ActivitySource is the type that help create and start Activity and allow listening to the Activity events too.
+    public sealed class ActivitySource : IDisposable
+    {
+        /// Construct an ActivitySource object with the input name and component version
+        public ActivitySource(string name, Version version);
+
+        /// Returns the ActivitySource name.
+        public string Name { get; }
+
+        /// Returns the ActivitySource version.
+        public Version Version { get; }
+
+        /// Creates a new Activity object if there is any listener to the Activity, returns null otherwise.
+        public Activity? StartActivity(string name, ActivityKind kind = ActivityKind.Internal);
+
+        public Activity? StartActivity(string name, ActivityKind kind, ActivityContext context, IEnumerable<KeyValuePair<string, string?>>? tags = null, IEnumerable<ActivityLink>? links = null, DateTimeOffset startTime = default);
+
+        /// Dispose the ActivitySource object and remove the current instance from the global list. empty the listeners list too.
+        public void Dispose();
+
+        /// Add a listener to the Activity starting and stopping events.
+        public IDisposable AddActivityListener(
+                                Func<ActivitySource, string, ActivityKind, ActivityContext, IEnumerable<KeyValuePair<string, string?>>?, IEnumerable<ActivityLink>?, ActivityCreationOptions> getActivityCreationOption,
+                                Action<Activity> onActivityStarted,
+                                Action<Activity> onActivityStopped);
+
+        /// Add a listener to listen to ActivitySource creation events.
+        public static IDisposable AddListener(Action<ActivitySource> OnSourceAttaching);
+    }
+```
+
+### Activity
+
+```C#
+namespace System.Diagnostics
+{
+    public partial class Activity : IDisposable // added IDisposable
+    {
+        ....
+        public ActivityKind Kind { get; set; } = ActivityKind.Internal;
```

Comment:
I think this should be immutable after creation, lets remove set

Reply:
I was not sure but will do what you have suggested.

Reply:
done